conditional-rendering.md ru (#844)
reactjs/ru.react.dev: src/content/learn/conditional-rendering.md

---
title: Условный рендеринг
---

<Intro>

> Вашим компонентам нужно часто отображать различные вещи в зависимости от различных условий. В React вы можете реденрить JSX в зависимости от его условий, используя JavaScript операторы. Такие, как `if`, `&&` и `? :`

Reviewer: забыли "н" в слове "рендерить"

Author: > забыли "н" в слове "рендерить"

Спасибо, исправил.

</Intro>

<YouWillLearn>

- Как вернуть разный JSX, в зависимости от его условия

> - Как условно включить или исключить фрагмент JSX

Reviewer 2: Как по условию добавить или убрать часть JSX

Author: я добавил это предложение, но разве первый вариант для русского человек не звучит лучше? меня просто "условно" немного напрягает и "в зависимости от условия" звучало бы лучше, не?

Reviewer 2: > в зависимости от **его** условия

Тут просто не понятно из контекста про чьи условия идет речь. Мне тоже не нравится "условно", можно "по условию" или "в зависимости от ситуации", например так:

- Как вернуть разный JSX, в зависимости от ситуации
- Как по условию добавить или убрать часть JSX

Author: > > в зависимости от **его** условия
> 
> Тут просто не понятно из контекста про чьи условия идет речь, в оригинале также нет указания на **чьи** то условия. Поэтому, возможно, и переводили как "условный рендеринг", но мне тоже не нравится "условно", можно "по условию" или "в зависимости от ситуации", например так:
> 
> * Как вернуть разный JSX, в зависимости от ситуации
> * Как по условию добавить или убрать часть JSX

Сделал правку, благодарю!

> - Общий условный синтаксис, который вы встретите в кодовой базе React.

Reviewer 2: ```suggestion
- Часто встречающиеся конструкции, которые вы найдете в проектах на React  
```

</YouWillLearn>

> ## Условно возвращаемый JSX {/*conditionally-returning-jsx*/}

Author: ```suggestion
## Условный возврат JSX {/*conditionally-returning-jsx*/}
```

> Допустим, у вас есть `PackingList` компонент, который рендерит несколько `Item`, которые могут быть обозначены, как упакованные или неупавкованные:

Reviewer 2: 2 раза подряд "который", кажется, можно поправить. 

```suggestion
Допустим, у вас есть компонент `PackingList`, который рендерит несколько компонентов `Item`, они могут быть обозначены, как упакованные или нет:
```

В английском нормально просто написать "several `Item`s", потому что компонент назван по английски, в русском могло быть что-то типа такого "который рендерит несколько `Вещей`". Но т.к. мы тоже называем компоненты по английски, то стоит добавить "компонентов", чтобы фраза не звучала обрывочно

```

</Sandpack>
                  

> Заметь, что некоторые `Item` компоненты имеют свой `isPacked` проп, который `true` вместо `false`. Вы хотите добавить галочку (✔) к упакованным вещам, если if `isPacked={true}`.

Reviewer 2: > Заметь
Т.к. мы с читателем на "вы", а "заметьте" сложно выговаривается, то предлагаю "Обратите внимание"

```suggestion
Обратите внимание, что некоторые компоненты `Item` имеют свойство `isPacked`, со значением `true` вместо `false`. Допустим, вы хотите добавить (✔) к упакованным вещам, если if `isPacked={true}`.
```
или
```suggestion
Обратите внимание, что некоторые компоненты `Item` имеют свойство `isPacked`, со значением `true` вместо `false`. Вам необходимо добавить (✔) к упакованным вещам, если if `isPacked={true}`.
```

> Вы можете писать это как [`if`/`else` условие](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Statements/if...) таким образом:

Reviewer 2: ```suggestion
Можно реализовать это с помощью [управляющей конструкции `if`/`else`](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Statements/if...) таким образом:
```

Reviewer 2: statement это все-таки управляющая конструкция, ну или инструкция, как переведено на прекрепленном здесь же mdn

);
```

Вы можете читать это как *"if `isPacked` это true, тогда (`?`) рендерим `name + ' ✔'`, в противном случае (`:`) рендерю `name`"*.

<DeepDive>

> #### Являются ли эти два примера полностью эквивалентными? {/*are-these-two-examples-fully-equivalent*/}

Author: Кажется, что первый вариант передает лучше ключевой смысл, не?

> #### Являются ли эти два примера полностью эквивалентными? {/*are-these-two-examples-fully-equivalent*/}

Reviewer 3: Смысл одинаковый, разница в лишнем слове в заголовке

> Если вы знакомы с объектно-ориентированным программированием, вы можете предположить, что два приведенных выше примера немного отличаются друг от друга, поскольку один из них может создавать два разных "экземпляра" `<li>`. Но JSX-элементы не являются "экземплярами", потому что они не хранят никакого внутреннего состояния и не являются реальными DOM-узлами . Это лёгкие описания, как чертежи. На самом деле эти два примера *совершенно эквивалентны*. В [Сохранение и сброс состояния](/learn/preserving-and-resetting-state) подробно рассказывается о том, как это работает.

Reviewer 3: ```suggestion
Если вы знакомы с объектно-ориентированным программированием, вы можете предположить, что два приведенных выше примера немного отличаются друг от друга, поскольку один из них может создавать два разных "экземпляра" `<li>`. Но JSX-элементы не являются "экземплярами", потому что они не хранят никакого внутреннего состояния и не являются реальными DOM-узлами. Это лёгкие описания, как чертежи. На самом деле эти два примера *совершенно эквивалентны*. В [Сохранение и сброс состояния](/learn/preserving-and-resetting-state) подробно рассказывается о том, как это работает.
```
У меня в предложенном варианте была ошибка с лишним пробелом. Исправляю.

Например, распространённой ошибкой является написание кода вида `messageCount && <p>Новые сообщения</p>`. Легко предположить, что ничего не отрендерено, когда `messageCount` равно `0`, но на самом деле будет рендериться `0`!

Чтобы исправить это, сделайте левую часть булевым значением: `messageCount > 0 && <p>Новые сообщения</p>`.

</Pitfall>

### Условное присвоение JSX к переменной {/*conditionally-assigning-jsx-to-a-variable*/}

Когда сокращения мешают написанию понятного кода, то попробуйте использовать `if` оператор и переменную. Вы можете переназначать переменные, объявленные с помощью [`let`](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Statements/let), поэтому начните с предоставления содержимого по умолчанию, которое вы хотите отобразить, например, name:

```js
let itemContent = name;
```

Используйте `if` оператор, чтобы переназначить JSX-выражение `itemContent`, если `isPacked` равно `true`:

```js
if (isPacked) {
  itemContent = name + ' ✔';
}
```

> [Фигурные скобки открывают "окно в мир JavaScript".](/learn/javascript-in-jsx-with-curly-braces#using-curly-braces-a-window-into-the-javascript-world) Вставьте переменную с фигурными скобками в возвращаемое дерево JSX, вложив ранее вычисленное выражение внутрь JSX:

Reviewer 3: ```suggestion
[Фигурные скобки открывают "окно в мир JavaScript".](/learn/javascript-in-jsx-with-curly-braces#using-curly-braces-a-window-into-the-javascript-world) Вставьте переменную с помощью фигурных скобок в возвращаемое дерево JSX, вложив ранее вычисленное выражение внутрь JSX:
```
Слегка перефразировал, сейчас более понятно причём тут фигурные скобки.

</Sandpack>

Помните, что вы должны писать `importance > 0 && ...` а не `importance && ...` и если `importance` это `0`, то `0` не зарендерится как результат!

> В этом решении, два раздельных условия были использованы, чтобы вставить пробел между именем и меткой важности. В качестве альтернативы можно использовать фрагмент с ведущим пробелом: `importance > 0 && <> <i>...</i></>` или добавить пробел сразу внутри тега `<i>`: `importance > 0 && <i> ...</i>`.

Reviewer 3: ```suggestion
В этом решении используются два отдельных условия для вставки пробела между именем и меткой важности. Кроме того, можно использовать фрагмент с ведущим пробелом: `importance > 0 && <> <i>...</i></>` или добавить пробел сразу внутри тега `<i>`: `importance > 0 && <i> ...</i>`.
```
Перефразировал и избавился от запятых ненужных.

</Solution>

> #### Перепешите тернарный оператор `? :` на `if` с переменной {/*refactor-a-series-of---to-if-and-variables*/}

Reviewer 3: ```suggestion
#### Перепишите тернарный оператор `? :` на `if` и переменные {/*refactor-a-series-of---to-if-and-variables*/}
```
Орфографию поправил, плюс в оригинале множественное число

> Этот `Drink` компонент использует серию `? :` условий, чтобы показать разную информацию, которая зависит от `name` пропа, который `"tea"` или `"coffee"`. Проблема в том, что информация о каждом напитке распределена по нескольким условиям. Перепешите этот код так, чтобы использовать одно `if` условие вместо трёх `? :` условий.

Reviewer 3: ```suggestion
Компонент `Drink` использует серию условий `? :` для отображения разной информацию в зависимости от пропа `name` (может быть `"tea"` или `"coffee"`). Проблема заключается в том, что информация о каждом напитке разбросана по нескольким условиям. Перепишите этот код так, используя один `if` вместо трёх условий `? :`.
```
Перефразировал, исправил ошибки

</Sandpack>

> После рефакторинга кода с использованием `if` у вас есть идеи, как его упростить?

Reviewer 3: ```suggestion
После рефакторинга кода с использованием `if`, у вас есть ещё идеи о том, как упростить его?
```
Перефразировал и добавил нужную запятую.

</Sandpack>

> Здесь информация о каждом напитке сгруппирована вместе, а не распределена по нескольким условиям. Это облегчает добавление новых напитков в будущем.

Reviewer 3: ```suggestion
Здесь информация о каждом напитке сгруппирована вместе, а не разбросана по нескольким условиям. Это облегчает добавление новых напитков в будущем.
```
Перефразировал, "разбросана", имхо, лучше подходит здесь.

> Другим решением может быть полное удаление условий путем перемещения информации в объекты:

Reviewer 3: ```suggestion
Другим решением будет удалить условия, переместив информацию в объекты:
```
Перефразировал и упростил, читать стало проще
